File: pkg/docs/slides/pptxworker/render_thumbs.go
package pptxworker

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os/exec"
	"time"
)

// ThumbProtocolVersion pins the request/response contract shared with
// render_thumbs.mjs (pptx-glimpse). Independent of the import worker protocol.
const ThumbProtocolVersion = 1

// ThumbRequest asks the thumbnail worker to rasterize every slide of a PPTX.
// Width is the output width in pixels; height follows the slide aspect ratio.
// Zero Width uses the worker default (480).
type ThumbRequest struct {
	ProtocolVersion int    `json:"protocolVersion"`
	PPTXBase64      string `json:"pptxBase64"`
	Width           int    `json:"width,omitempty"`
}

// ThumbSlide is one rendered slide. PNGBase64 is empty when that slide failed;
// the failure reason is in Warnings.
type ThumbSlide struct {
	SlideNumber int      `json:"slideNumber"`
	PNGBase64   string   `json:"pngBase64"`
	Width       int      `json:"width"`
	Height      int      `json:"height"`
	Warnings    []string `json:"warnings,omitempty"`
}

// ThumbResponse is the single JSON message render_thumbs.mjs writes to stdout.
type ThumbResponse struct {
	ProtocolVersion int          `json:"protocolVersion"`
	Slides          []ThumbSlide `json:"slides,omitempty"`
	Warnings        []string     `json:"warnings,omitempty"`
	Error           string       `json:"error,omitempty"`
}

// ThumbRunner invokes render_thumbs.mjs through Node. WorkingDir must be the
// directory that contains node_modules/node-pptx-png (web/).
type ThumbRunner struct {
	NodePath   string
	WorkingDir string
	ScriptPath string
	Timeout    time.Duration
}

// Run renders every slide. A worker-level error is returned as a Go error.
// Per-slide failures stay in ThumbResponse.Slides so the caller can still
// show the slides that rendered.
func (r ThumbRunner) Run(ctx context.Context, req ThumbRequest) (ThumbResponse, error) {
	if req.ProtocolVersion == 0 {
		req.ProtocolVersion = ThumbProtocolVersion
	}
	if req.ProtocolVersion != ThumbProtocolVersion {
		return ThumbResponse{}, fmt.Errorf("unsupported pptx thumbnail protocol %d", req.ProtocolVersion)
	}
	if req.PPTXBase64 == "" {
		return ThumbResponse{}, fmt.Errorf("pptx thumbnail worker: empty pptx")
	}
	payload, err := json.Marshal(req)
	if err != nil {
		return ThumbResponse{}, fmt.Errorf("marshal pptx thumbnail request: %w", err)
	}
	timeout := r.Timeout
	if timeout <= 0 {
		timeout = 120 * time.Second
	}
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	node := r.NodePath
	if node == "" {
		node = "node"
	}
	cmd := exec.CommandContext(ctx, node, r.ScriptPath)
	cmd.Dir = r.WorkingDir
	cmd.Stdin = bytes.NewReader(payload)
	var stdout, stderr bytes.Buffer
	cmd.Stdout, cmd.Stderr = &stdout, &stderr
	if err := cmd.Run(); err != nil {
		if ctx.Err() != nil {
			return ThumbResponse{}, fmt.Errorf("pptx thumbnail worker timeout: %w", ctx.Err())
		}
		return ThumbResponse{}, fmt.Errorf("pptx thumbnail worker: %w: %s", err, stderr.String())
	}
	var response ThumbResponse
	if err := json.Unmarshal(stdout.Bytes(), &response); err != nil {
		return ThumbResponse{}, fmt.Errorf("decode pptx thumbnail worker response: %w: %s", err, stderr.String())
	}
	if response.ProtocolVersion != ThumbProtocolVersion {
		return ThumbResponse{}, fmt.Errorf("pptx thumbnail worker returned protocol %d", response.ProtocolVersion)
	}
	if response.Error != "" {
		return ThumbResponse{}, fmt.Errorf("pptx thumbnail worker: %s", response.Error)
	}
	return response, nil
}

// DecodePNG returns the raw PNG bytes for a rendered slide.
func (s ThumbSlide) DecodePNG() ([]byte, error) {
	if s.PNGBase64 == "" {
		return nil, fmt.Errorf("slide %d has no png", s.SlideNumber)
	}
	raw, err := base64.StdEncoding.DecodeString(s.PNGBase64)
	if err != nil {
		return nil, fmt.Errorf("slide %d png: %w", s.SlideNumber, err)
	}
	return raw, nil
}

File: pkg/docs/slides/pptxworker/render_thumbs.mjs
// PPTX → PNG thumbnail worker for slides-diag.
//
// Renders the original .pptx with pptx-glimpse (in-process SVG/PNG, no LibreOffice),
// not the import IR. One JSON request on stdin, one JSON response on stdout.
// Spawn with cwd = web/ so node_modules resolve.
//
// Request:  { "protocolVersion": 1, "pptxBase64": "...", "width": 480 }
// Response: { "protocolVersion": 1, "slides": [{ "slideNumber": 1, "pngBase64": "...", "width": 480, "height": 270, "warnings": [] }], "warnings": [], "error": "" }

import { createRequire } from 'node:module'
import { pathToFileURL } from 'node:url'

const PROTOCOL = 1

const readStdin = async () => {
  const chunks = []
  for await (const chunk of process.stdin) chunks.push(chunk)
  return Buffer.concat(chunks).toString('utf8')
}

const fail = (message) => {
  process.stdout.write(JSON.stringify({ protocolVersion: PROTOCOL, error: String(message) }))
  process.exit(0)
}

const warnText = (d) => {
  if (!d) return ''
  if (typeof d === 'string') return d
  const code = d.code || d.kind || d.severity || ''
  const msg = d.message || d.text || JSON.stringify(d)
  return code ? `[${code}] ${msg}` : String(msg)
}

const main = async () => {
  let req
  try {
    req = JSON.parse(await readStdin() || '{}')
  } catch (err) {
    fail(`invalid JSON request: ${err.message}`)
    return
  }
  if (req.protocolVersion && req.protocolVersion !== PROTOCOL) {
    fail(`unsupported protocol ${req.protocolVersion}`)
    return
  }
  if (!req.pptxBase64) {
    fail('pptxBase64 is required')
    return
  }
  let bytes
  try {
    bytes = Buffer.from(req.pptxBase64, 'base64')
  } catch (err) {
    fail(`pptxBase64 decode: ${err.message}`)
    return
  }
  const width = Number(req.width) > 0 ? Number(req.width) : 480

  let convertPptxToPng
  let result
  try {
    ;({ convertPptxToPng } = await import(pathToFileURL(createRequire(process.cwd() + '/package.json').resolve('pptx-glimpse')).href))
  } catch (err) {
    fail(`pptx-glimpse is not installed in the worker working directory (web/node_modules): ${err.message}`)
    return
  }
  try {
    result = await convertPptxToPng(bytes, { width })
  } catch (err) {
    try {
      result = await convertPptxToPng(bytes)
    } catch (err2) {
      fail(`render: ${err2.message || err.message}`)
      return
    }
  }

  const rendered = result?.slides || result || []
  const warnings = (result?.diagnostics || []).map(warnText).filter(Boolean)
  const slides = []
  for (let i = 0; i < rendered.length; i++) {
    const slide = rendered[i]
    const png = slide?.png || slide?.image || slide?.data
    if (!png) {
      warnings.push(`slide ${i + 1}: empty png`)
      slides.push({ slideNumber: i + 1, pngBase64: '', width: 0, height: 0, warnings: ['empty png'] })
      continue
    }
    const buf = Buffer.isBuffer(png) ? png : Buffer.from(png)
    slides.push({
      slideNumber: slide.slideNumber || slide.index + 1 || i + 1,
      pngBase64: buf.toString('base64'),
      width: slide.width || width,
      height: slide.height || 0,
      warnings: (slide.diagnostics || slide.warnings || []).map(warnText).filter(Boolean),
    })
  }

  process.stdout.write(JSON.stringify({ protocolVersion: PROTOCOL, slides, warnings }))
}

main().catch((err) => fail(err?.stack || err?.message || String(err)))
